src/main/java/com/google/sps/servlets/DeleteCardServlet.java

```diff
@@ -0,0 +1,124 @@
+package com.google.sps.servlets;
+
+import static com.google.appengine.api.datastore.TransactionOptions.Builder.*;
+
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import com.google.appengine.api.datastore.Transaction;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.KeyFactory;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+
+import com.google.appengine.api.taskqueue.Queue;
+import com.google.appengine.api.taskqueue.QueueFactory;
+import com.google.appengine.api.taskqueue.TaskOptions;
+
+import com.google.appengine.api.blobstore.BlobstoreFailureException;
+import com.google.appengine.api.blobstore.BlobKey;
+import com.google.appengine.api.blobstore.BlobstoreService;
+import com.google.appengine.api.blobstore.BlobstoreServiceFactory;
+
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+import com.google.gson.Gson;
+import com.google.sps.tool.ResponseSerializer;
+
+@WebServlet("/deletecard")
+public class DeleteCardServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+
+    if (userService.isUserLoggedIn()) {
+      DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+      String cardKey = request.getParameter("cardKey");
+      Entity card = getExistingCardInDatastore(datastore, cardKey);
+
+      if (card == null) {
+        String jsonErrorInfo = ResponseSerializer.getErrorJson("Cannot delete Card at the moment");
+        response.setContentType("application/json;");
+        response.getWriter().println(new Gson().toJson(jsonErrorInfo));
+      } else {
+        deleteBlob(card);
+        deleteCard(card);
+      }
+    }
+  }
+
+  private Entity getExistingCardInDatastore(DatastoreService datastore, String cardKey)
+      throws IOException {
+    try {
+      return datastore.get(KeyFactory.stringToKey(cardKey));
+    } catch (EntityNotFoundException e) {
+      return null;
+    }
+  }
+
+  private void deleteBlob(Entity card) throws IOException {
```

Comment:
This shares a lot of duplicate code with #13. Since these are likely doing "the same thing," you should find a way to deduplicate.

Reply:
done